2-Draft-Accepted/RFCxxxx-PowerShellGet-3.0.md

```diff
@@ -0,0 +1,383 @@
+`Find-PSResource` (see below) works against a local cache.
+This will also enable changes in PowerShell to use `Find-PSResource -Type Command` to look
+in this cache when it can't find a command and suggest to the user the module to install to
+get that command.
+This will be a local json file containing only the latest version of each module.
+The cache will be stored in the user path.
+There is no system cache that is shared.
+A system wide cache would require elevation or sudo to create and update preventing
+it from being useful.
+
+Example cache entry:
+
+```json
+{
+  "name": "This is my module",
+  "exportedFunctions": [
+    "Get-One",
+    "Set-Two"
+  ],
+  "version": "1.0.0.0",
+  "type": "module",
+  "tags": [
+    "Linux",
+    "PSEdition_Core",
+    "PSEdition_Desktop",
+    "AzureAutomation"
+  ]
+}
+```
+
+An example cache with 5000 resources (approximately the number of unique packages
+currently published to PowerShellGallery.com) is ~700KB in compressed json form.
+
+The cache would have both latest stable and latest prerelease versions of resources.
+
+>[!NOTE]
+>Need to experiment if it makes sense to have a single cache file for all repositories
+>or a different cache per repository for size and perf reasons.
+>Perf tests will determine if the cache needs to be in binary form.
+
+### Automatic updating of the cache
+
+On any operation that reaches out to a repository, a REST API will be called to
+see if the hash of the cache matches the current cache and if not, a new one
+is downloaded.
+
+>[!NOTE]
+>In the implementation we'll need to handle the case where the module attempts
+>to update the cache while it is being read.
+
+If the repository doesn't support this new API, it falls back to current behavior
+in PSGet v2.
+This means that there is no local cache of that repository and operations will
+always connect to that repository.
+
+### Repository management
+
+`Register-PSResourceRepository` will allow for registering additional repositories.
+A `-Default` switch enables registering PSGallery should it be accidentally removed.
+The `-URL` will accept the HTTP address without the need to specify `/api/v3` as
+that will be assumed and discovered at runtime (trying v3 first, then falling
+back to v2, then the literal URL).
+Support for local filesystem repositories must be maintained.
+A `-Trusted` switch indicates whether to prompt the user when installing resources
+from that repository.
+By default, if this switch is not specified, the repository is untrusted.
+A `-Repositories` parameter will accept an array of hashtables equivalent to
+the parameter names (like splatting).
+
+```powershell
+Register-PSResourceRepository -Repositories @(
+  @{ URL = "https://one.com"; Trusted = $true }
+  @{ URL = "https://powershellgallery.com"; Trusted = $true; Default = $true }
+)
+```
+
+A `-Name` parameter allows for setting a friendly name.
+
+A `-Default` switch will set one repository as the default (when not specified
```

Comment:
I really think the list of repositories should just have a priority weighting so I could assign them a number, and they'd be consulted in order and the first hit would get used.

1. What's the point of a "default"?
2. In what situation will "only" the default be searched?
3. Will you ever install an older version just because the newer one is in a feed that's not "default"?

Reply:
`apt` uses the order defined in `sources.list` to indicate priority.  I don't want users to edit a file manually.  I suppose we could have a property called `Priority` where lower is higher priority.  Default value is 50 if not specified.  Repos with the same priority will be searched in the way they are registered.  PSGallery will have priority 25.

Reply:
Priority property in the sources list is what chocolatey uses so wouldn't be totally alien to people if they're defining their own sources list anyway.

Reply:
For organizations using an internal repository as the primary source for modules. I would expect to see user registered repositories to have a higher default priority than PSGallery. 

@SteveL-MSFT What thought behind PSGallery having a higher default priority than user registered repositories?

Reply:
Makes sense for user registered ones to have higher priority.  Will invert this.